usr/src/uts/armv8/os/fakebop.c

```diff
@@ -592,54 +970,485 @@ bop_panic(const char *fmt, ...)
 }
 
-static int
-parse_value(char *p, uint64_t *retval)
+/*
+ * XXXARM: We should consider having bop_sysp again, which could
+ * call out to functions in inetboot/shim code during early boot.
+ */
+
+static void
+build_firmware_properties_fdt(const void *fdtp)
 {
-	int adjust = 0;
-	uint64_t tmp = 0;
-	int digit;
-	int radix = 10;
+	int property;
+	int node;
 
-	*retval = 0;
-	if (*p == '-' || *p == '~')
-		adjust = *p++;
+	node = fdt_subnode_offset(fdtp, 0, "chosen");
+	if (node < 0)
+		return;
 
-	if (*p == '0') {
-		++p;
-		if (*p == 0)
-			return (0);
-		if (*p == 'x' || *p == 'X') {
-			radix = 16;
-			++p;
+	fdt_for_each_property_offset(property, fdtp, node) {
+		const char *pn;
+		const void *pv;
+		int pl;
+
+		pv = fdt_getprop_by_offset(fdtp, property, &pn, &pl);
+		if (pv == NULL)
+			continue;
+		if (do_bsys_getproplen(NULL, pn) >= 0)
+			continue;
+
+		if (strcmp(pn, "phandle") == 0 ||
+		    strcmp(pn, "impl-arch-name") == 0 ||
+		    strcmp(pn, "mfg-name") == 0 ||
+		    strcmp(pn, "__bootpath") == 0) {
+			continue;
+		}
+
+		if (pl == 0) {
+			bsetprop(DDI_PROP_TYPE_ANY, (char *)pn, strlen(pn),
+			    NULL, 0);
+		} else if (strcmp((char *)pn, "display-edif-block") == 0 ||
+		    strcmp((char *)pn, "kaslr-seed") == 0 ||
+		    strcmp((char *)pn, "rng-seed") == 0) {
+			bsetprop(DDI_PROP_TYPE_BYTE, (char *)pn, strlen(pn),
+			    (void *)pv, pl);
 		} else {
-			radix = 8;
-			++p;
+			bsetprop(DDI_PROP_TYPE_STRING, (char *)pn, strlen(pn),
+			    (void *)pv, pl);
 		}
 	}
-	while (*p) {
-		if ('0' <= *p && *p <= '9')
-			digit = *p - '0';
-		else if ('a' <= *p && *p <= 'f')
-			digit = 10 + *p - 'a';
-		else if ('A' <= *p && *p <= 'F')
-			digit = 10 + *p - 'A';
-		else
-			return (-1);
-		if (digit >= radix)
-			return (-1);
-		tmp = tmp * radix + digit;
-		++p;
+}
+
+/*
+ * Populate properties from firmware values.
+ */
+static void
+build_firmware_properties(struct xboot_info *xbp)
+{
+	if (xbp->bi_fdt != 0) {
+		const void *fdtp = (void *)xbp->bi_fdt;
+
+		if (fdt_check_header(fdtp) == 0) {
+			build_firmware_properties_fdt(fdtp);
+		}
 	}
-	if (adjust == '-')
-		tmp = -tmp;
-	else if (adjust == '~')
-		tmp = ~tmp;
-	*retval = tmp;
-	return (0);
 }
 
+/*
+ * Import boot environment module variables as properties, applying
+ * blocklist filter for variables we know we will not use.
+ *
+ * Since the environment can be relatively large, containing many variables
+ * used only for boot loader purposes, we will use a blocklist based filter.
+ * To keep the blocklist from growing too large, we use prefix based filtering.
+ * This is possible because in many cases, the loader variable names are
+ * using a structured layout.
+ *
+ * We will not overwrite already set properties.
+ *
+ * Note that the menu items in particular can contain characters not
+ * well-handled as bootparams, such as spaces, brackets, and the like, so that's
+ * another reason.
+ */
+static struct bop_blocklist {
```

Comment:
I guess this is from x86, and assumes what loader would do if we had it?

Reply:
Yes, from i86pc. I have a port of loader64.efi, which sets a lot of this stuff.

For that port, I'm working to straighten out how we can recognise and reason about ttya/ttyb etc. (since we don't have a "standard I/O port addressing" scheme to fall back on).